File: GolfIt/MovingFloor.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Text;
using System.Threading.Tasks;
using static System.Windows.Forms.VisualStyles.VisualStyleElement.Rebar;

namespace GolfIt
{
    public class MovingFloor : Obstacle
    {
        public Vector position;
        public Direction direction;
        private int cellSize;
        private float floorForce;
        private Point[] points;

        public MovingFloor(Vector position, Direction direction, int cellSize)
        {
            this.position = position;
            this.direction = direction;
            this.cellSize = cellSize;
            this.floorForce = 0.15f;

            points = new Point[6];

            CalculateArrowPoints();
        }

        public Vector DetectCollision(Ball ball)
        { 
            var velocity = new Vector(0, 0);

            if (!Collided(ball))
            {
                return velocity;
            }

            switch (direction)
            {
                case Direction.Up:
                    velocity.Y = -floorForce;
                    break;
                case Direction.Down:
                    velocity.Y = floorForce;
                    break;
                case Direction.Left:
                    velocity.X = -floorForce;
                    break;
                case Direction.Right:
                    velocity.X = floorForce;
                    break;
            }

            return velocity;
        }

        public bool Collided(Ball ball)
        {
            bool top = IsMovingFloor((int)(ball.position.X), (int)(ball.position.Y - ball.radius));
            bool bottom = IsMovingFloor((int)(ball.position.X), (int)(ball.position.Y + ball.radius));
            bool left = IsMovingFloor((int)(ball.position.X - ball.radius), (int)(ball.position.Y));
            bool right = IsMovingFloor((int)(ball.position.X + ball.radius), (int)(ball.position.Y));
            bool topLeft = IsMovingFloor((int)(ball.position.X - ball.radius), (int)(ball.position.Y - ball.radius));
            bool topRight = IsMovingFloor((int)(ball.position.X + ball.radius), (int)(ball.position.Y - ball.radius));
            bool bottomLeft = IsMovingFloor((int)(ball.position.X - ball.radius), (int)(ball.position.Y + ball.radius));
            bool bottomRight = IsMovingFloor((int)(ball.position.X + ball.radius), (int)(ball.position.Y + ball.radius));

            return top || bottom || left || right || topLeft || topRight || bottomLeft || bottomRight;
        }

        public bool IsMovingFloor(int x, int y)
        {
            int gridX = x / cellSize;
            int gridY = y / cellSize;


            return gridX == position.X && gridY == position.Y;
        }

        public void Render(Graphics g, PictureBox canvas)
        {
            g.FillRectangle(Brushes.Yellow, position.X * cellSize, position.Y * cellSize, cellSize, cellSize);
            g.FillPolygon(Brushes.Orange, points);
        }
        public void Update(Graphics g, PictureBox canvas, Map map, int cntT)
        {
            Render(g, canvas);
        }

        private Point[] CalculateArrowPoints()
        {
            int halfCell = cellSize / 2;
            int x = (int)position.X;
            int y = (int)position.Y;

            switch (direction)
            {
                case Direction.Up:
                    points[0] = new Point(x * cellSize, (y + 1) * cellSize);
                    points[1] = new Point(x * cellSize + halfCell, y * cellSize + halfCell);
                    points[2] = new Point((x + 1) * cellSize, (y + 1) * cellSize);
                    points[3] = new Point((x + 1) * cellSize, y * cellSize + halfCell);
                    points[4] = new Point(x * cellSize + halfCell, y * cellSize);
                    points[5] = new Point(x * cellSize, y * cellSize + halfCell);
                    break;
                case Direction.Down:
                    points[0] = new Point(x * cellSize, y * cellSize);
                    points[1] = new Point(x * cellSize + halfCell, y * cellSize + halfCell);
                    points[2] = new Point((x + 1) * cellSize, y * cellSize);
                    points[3] = new Point((x + 1) * cellSize, y * cellSize + halfCell);
                    points[4] = new Point(x * cellSize + halfCell, (y + 1) * cellSize);
                    points[5] = new Point(x * cellSize, y * cellSize + halfCell);
                    break;
                case Direction.Left:
                    points[0] = new Point((x + 1) * cellSize, y * cellSize);
                    points[1] = new Point(x * cellSize + halfCell, y * cellSize + halfCell);
                    points[2] = new Point((x + 1) * cellSize, (y + 1) * cellSize);
                    points[3] = new Point(x * cellSize + halfCell, (y + 1) * cellSize);
                    points[4] = new Point(x * cellSize, y * cellSize + halfCell);
                    points[5] = new Point(x * cellSize + halfCell, y * cellSize);
                    break;
                case Direction.Right:
                    points[0] = new Point(x * cellSize, y * cellSize);
                    points[1] = new Point(x * cellSize + halfCell, y * cellSize + halfCell);
                    points[2] = new Point(x * cellSize, (y + 1) * cellSize);
                    points[3] = new Point(x * cellSize + halfCell, (y + 1) * cellSize);
                    points[4] = new Point((x + 1) * cellSize, y * cellSize + halfCell);
                    points[5] = new Point(x * cellSize + halfCell, y * cellSize);
                    break;
            }

            return points;
        }

    }

    public enum Direction
    {
        Up, Down, Left, Right,
    }
}

File: golfit_rs/src/moving_floor.rs
use raylib::prelude::*;
use crate::vector::Vector2D;
use crate::obstacle::Obstacle;
use crate::map::Map;
use crate::ball::Ball;

#[derive(Debug, Clone, Copy)]
pub enum Direction {
    Up,
    Down,
    Left,
    Right,
}

pub struct MovingFloor {
    pub position: Vector2D,
    pub direction: Direction,
    cell_size: i32,
    floor_force: f32,
    points: [Vector2; 6],
}

impl MovingFloor {
    pub fn new(position: Vector2D, direction: Direction, cell_size: i32) -> Self {
        let mut moving_floor = Self {
            position,
            direction,
            cell_size,
            floor_force: 0.15,
            points: [Vector2::new(0.0, 0.0); 6],
        };
        moving_floor.calculate_arrow_points();
        moving_floor
    }

    pub fn collided(&self, ball: &Ball) -> bool {
        let top = self.is_moving_floor(ball.position.x as i32, (ball.position.y - ball.radius) as i32);
        let bottom = self.is_moving_floor(ball.position.x as i32, (ball.position.y + ball.radius) as i32);
        let left = self.is_moving_floor((ball.position.x - ball.radius) as i32, ball.position.y as i32);
        let right = self.is_moving_floor((ball.position.x + ball.radius) as i32, ball.position.y as i32);
        let top_left = self.is_moving_floor((ball.position.x - ball.radius) as i32, (ball.position.y - ball.radius) as i32);
        let top_right = self.is_moving_floor((ball.position.x + ball.radius) as i32, (ball.position.y - ball.radius) as i32);
        let bottom_left = self.is_moving_floor((ball.position.x - ball.radius) as i32, (ball.position.y + ball.radius) as i32);
        let bottom_right = self.is_moving_floor((ball.position.x + ball.radius) as i32, (ball.position.y + ball.radius) as i32);

        top || bottom || left || right || top_left || top_right || bottom_left || bottom_right
    }

    fn is_moving_floor(&self, x: i32, y: i32) -> bool {
        let grid_x = (x - 8) / self.cell_size; // Account for canvas offset
        let grid_y = (y - 34) / self.cell_size;
        
        // Check if this position matches our moving floor position
        let pos_x = (self.position.x as i32) / self.cell_size;
        let pos_y = (self.position.y as i32) / self.cell_size;
        
        grid_x == pos_x && grid_y == pos_y
    }

    fn calculate_arrow_points(&mut self) {
        let cell_f = self.cell_size as f32;
        let _half_cell = cell_f / 2.0;
        let quarter_cell = cell_f / 4.0;

        match self.direction {
            Direction::Up => {
                self.points = [
                    Vector2::new(self.position.x, self.position.y - quarter_cell),
                    Vector2::new(self.position.x - quarter_cell, self.position.y),
                    Vector2::new(self.position.x - quarter_cell / 2.0, self.position.y),
                    Vector2::new(self.position.x - quarter_cell / 2.0, self.position.y + quarter_cell),
                    Vector2::new(self.position.x + quarter_cell / 2.0, self.position.y + quarter_cell),
                    Vector2::new(self.position.x + quarter_cell / 2.0, self.position.y),
                ];
            }
            Direction::Down => {
                self.points = [
                    Vector2::new(self.position.x, self.position.y + quarter_cell),
                    Vector2::new(self.position.x - quarter_cell, self.position.y),
                    Vector2::new(self.position.x - quarter_cell / 2.0, self.position.y),
                    Vector2::new(self.position.x - quarter_cell / 2.0, self.position.y - quarter_cell),
                    Vector2::new(self.position.x + quarter_cell / 2.0, self.position.y - quarter_cell),
                    Vector2::new(self.position.x + quarter_cell / 2.0, self.position.y),
                ];
            }
            Direction::Left => {
                self.points = [
                    Vector2::new(self.position.x - quarter_cell, self.position.y),
                    Vector2::new(self.position.x, self.position.y - quarter_cell),
                    Vector2::new(self.position.x, self.position.y - quarter_cell / 2.0),
                    Vector2::new(self.position.x + quarter_cell, self.position.y - quarter_cell / 2.0),
                    Vector2::new(self.position.x + quarter_cell, self.position.y + quarter_cell / 2.0),
                    Vector2::new(self.position.x, self.position.y + quarter_cell / 2.0),
                ];
            }
            Direction::Right => {
                self.points = [
                    Vector2::new(self.position.x + quarter_cell, self.position.y),
                    Vector2::new(self.position.x, self.position.y - quarter_cell),
                    Vector2::new(self.position.x, self.position.y - quarter_cell / 2.0),
                    Vector2::new(self.position.x - quarter_cell, self.position.y - quarter_cell / 2.0),
                    Vector2::new(self.position.x - quarter_cell, self.position.y + quarter_cell / 2.0),
                    Vector2::new(self.position.x, self.position.y + quarter_cell / 2.0),
                ];
            }
        }
    }
}

impl Obstacle for MovingFloor {
    fn detect_collision(&self, ball: &Ball) -> Vector2D {
        let mut velocity = Vector2D::zero();

        if !self.collided(ball) {
            return velocity;
        }

        match self.direction {
            Direction::Up => velocity.y = -self.floor_force,
            Direction::Down => velocity.y = self.floor_force,
            Direction::Left => velocity.x = -self.floor_force,
            Direction::Right => velocity.x = self.floor_force,
        }

        velocity
    }

    fn render(&self, d: &mut RaylibDrawHandle) {
        // Draw the arrow shape
        for i in 0..self.points.len() {
            let next_i = (i + 1) % self.points.len();
            d.draw_line_v(self.points[i], self.points[next_i], Color::BLUE);
        }
        
        // Fill the arrow (simple approach)
        let center = Vector2::new(self.position.x, self.position.y);
        d.draw_circle_v(center, (self.cell_size / 4) as f32, Color::BLUE);
    }

    fn update(&mut self, d: &mut RaylibDrawHandle, _map: &Map, _cnt_t: u32) {
        self.render(d);
    }
}
